### PR TITLE
Solutions for Chapter 4

### PR DESCRIPTION
**Pull request by visar**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

**Comment by visar** [comment]
cc @vrom911 @chshersh

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,14 +478,14 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (Trap e) <*> _ = Trap e
+    (Reward f) <*> a = fmap f a
```
Comment (vrom911) [inline src/Chapter4.hs:485]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -488,6 +494,19 @@ Implement the 'Applicative' instance for our 'List' type.
   may also need to implement a few useful helper functions for our List
   type.
 -}
+append :: List a -> List a -> List a
+append Empty x = x
+append x Empty = x
```
Comment (vrom911) [inline src/Chapter4.hs:499]: Great catch! 🚤

File: src/Chapter4.hs [hunk 3]
```diff
@@ -629,7 +656,7 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+a `andM` b = a >>= (\c -> if c then b else pure False)
```
Comment (vrom911) [inline src/Chapter4.hs:659]: Yeap, this is correct lazy implementation! 💪🏼

**Review by vrom911 — APPROVED** [review]
:tada: Fantastic work! Clear and neat solutions ✨ 
Thank you so much for taking this course! I hope you had fun :man_dancing: 💃🏼 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face:

**Pull request merged**